### PR TITLE
Document forms with Symfony

### PR DESCRIPTION
**Pull request by thomasnares**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 8.x
| Description?  | Document forms with Symfony

Left todo: 

- make a demo module implementing `action<FormName>FormDataProviderDefaultData` and `action<FormName>FormDataProviderData` to confirm behavior
- improve internal linking between all discussed hooks in this page & the hook pages, link between example modules and this page, link between CQRS page and this page

Diagrams for easier review: 

Screenshot 2023-10-24 at 09 26 40

Screenshot 2023-10-24 at 09 26 48

Screenshot 2023-10-24 at 09 27 34

### CODE REVIEW

**Review by tleon — APPROVED** [review]
Could you try to squash some of those commits ? I think you can get rid of the "wip" commits ;)

**Comment by thomasnares** [comment]
> Could you try to squash some of those commits ? I think you can get rid of the "wip" commits ;)

Thanks, done 👍

**Review by matthieu-rolland — COMMENTED** [review]
there's a rebase to be made and some conflicts to be dealt with ;)

**Comment by thomasnares** [comment]
> there's a rebase to be made and some conflicts to be dealt with ;)

All done 👍

File: development/components/form/_index.md [hunk 1]
```diff
@@ -1,8 +1,276 @@
...
+useMermaid: true
 ---
 
-# The Form component
+# Forms with Symfony
 
-{{% children /%}}
+## Introduction
+
+With Symfony, `Forms` are built, rendered and handled in separate layers. Several concepts are involved, such as form building, providing data to forms, validating data, persisting data submited by forms, rendering forms, ...
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
Symfony's `Forms` system separates the construction, rendering, and processing of forms. This involves multiple key concepts, including form creation, data provision, validation, data persistence from submitted forms, and form rendering.
```

File: development/components/form/_index.md [hunk 2]
```diff
@@ -1,8 +1,276 @@
...
+        end
+    end
+    G-.->ZZ(Redirect)
+    style F fill:#f9f,stroke:#333,stroke-width:2px
+    style Core fill:#fbfbea
+</div>
+
+### Controllers
+
+`Controllers` in the back office receives http requests, and when handling requests for creating or editing an entity, they instanciate and uses several concepts.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
`Controllers` in the back office receive HTTP requests. When handling requests for creating or editing an entity, they instantiate and utilize several concepts.
```

File: development/components/form/_index.md [hunk 3]
```diff
@@ -1,8 +1,276 @@
...
+
+### FormBuilder
+
+In most situations, the first component instanciated by the `Controller` is the `FormBuilder`.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
In most scenarios, the Controller initially instantiates the FormBuilder. This component constructs the necessary Form, incorporating all Form Types. It also retrieves data from the FormDataProvider, ensuring that the form is populated with the appropriate data.
```

File: development/components/form/_index.md [hunk 4]
```diff
@@ -1,8 +1,276 @@
...
+`Controllers` in the back office receives http requests, and when handling requests for creating or editing an entity, they instanciate and uses several concepts.
+
+### FormBuilder
+
+In most situations, the first component instanciated by the `Controller` is the `FormBuilder`. 
+The `FormBuilder` will build the required `Form` with all `Form Types`, and retrieve data from the `FormDataProvider`. 
+
+### FormDataProvider
+
+The `FormDataProvider` is in charge of retrieving data from the `Database` to fill the form (when editing), and provide default data (when creating and editing).
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
The FormDataProvider is responsible for retrieving data from the Database. This data is used to populate the form when editing an existing entity. Additionally, it provides default values for the form fields, both when creating a new entity and when editing an existing one.
```

File: development/components/form/_index.md (hunk 2 above)
Comment (kpodemski) [reply development/components/form/_index.md]: I'm not sure about adding `Highlights` to the names that are not names of the classes, this is why I decided to get rid of them later :D

File: development/components/form/_index.md [hunk 5]
```diff
@@ -1,8 +1,276 @@
...
+In most situations, the first component instanciated by the `Controller` is the `FormBuilder`. 
+The `FormBuilder` will build the required `Form` with all `Form Types`, and retrieve data from the `FormDataProvider`. 
+
+### FormDataProvider
+
+The `FormDataProvider` is in charge of retrieving data from the `Database` to fill the form (when editing), and provide default data (when creating and editing). 
+
+### Form Types
+
+Developers can already use a large list of field types ([see Symfony types](https://symfony.com/doc/4.4/reference/forms/types.html)) that comes from the Symfony framework. In addition to that, PrestaShop adds more reusable field types that developers can use.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
Developers have access to a wide range of field types ([see Symfony types](https://symfony.com/doc/4.4/reference/forms/types.html)) that come from the Symfony framework. Additionally, PrestaShop enhances this selection with its own reusable field types.
```

File: development/components/form/_index.md [hunk 6]
```diff
@@ -1,8 +1,276 @@
...
+The `FormHandler` is responsible of:
+
+- extracting the form data, 
+- triggering the `actionBeforeCreate<FormName>FormHandler`, `actionBeforeUpdate<FormName>FormHandler`, `actionAfterCreate<FormName>FormHandler` and `actionAfterUpdate<FormName>FormHandler` hooks
+- calling the `FormDataHandler`
+- returning a `FormHandlerResult` 
+
+### FormDataHandler
+
+The `FormDataHandler` is responsible of persisting the data to `Database`, by sending a `CQRS command` to the `command bus`.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
The `FormDataHandler` is tasked with persisting data to the Database. It accomplishes this by dispatching a CQRS command to the command bus, ensuring efficient and effective data handling.
```

File: development/components/form/_index.md [hunk 7]
```diff
@@ -1,8 +1,276 @@
...
+- triggering the `actionBeforeCreate<FormName>FormHandler`, `actionBeforeUpdate<FormName>FormHandler`, `actionAfterCreate<FormName>FormHandler` and `actionAfterUpdate<FormName>FormHandler` hooks
+- calling the `FormDataHandler`
+- returning a `FormHandlerResult` 
+
+### FormDataHandler
+
+The `FormDataHandler` is responsible of persisting the data to `Database`, by sending a `CQRS command` to the `command bus`. 
+
+{{% notice note %}}
+This page is about Migrated `Forms` with `Symfony` and `CQRS pattern`.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
This page focuses on migrated Forms utilizing Symfony and the CQRS pattern.
```

File: development/components/form/_index.md [hunk 8]
```diff
@@ -1,8 +1,276 @@
...
+- returning a `FormHandlerResult` 
+
+### FormDataHandler
+
+The `FormDataHandler` is responsible of persisting the data to `Database`, by sending a `CQRS command` to the `command bus`. 
+
+{{% notice note %}}
+This page is about Migrated `Forms` with `Symfony` and `CQRS pattern`. 
+
+There are several ways to persist data, such as using [Doctrine entities]({{<relref "/8/modules/concepts/doctrine/#using-doctrine">}}), using [ObjectModel entities]({{<relref "/8/development/components/database/objectmodel">}}), or using [CQRS concepts]({{<relref "/8/development/architecture/domain/cqrs">}}) or not.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
It's important to note that data can be persisted in several ways, including: [Doctrine entities]({{<relref "/8/modules/concepts/doctrine/#using-doctrine">}}), using [ObjectModel entities]({{<relref "/8/development/components/database/objectmodel">}}), or using [CQRS concepts]({{<relref "/8/development/architecture/domain/cqrs">}}).
```

File: development/components/form/_index.md [hunk 9]
```diff
@@ -1,8 +1,276 @@
...
+{{% /notice %}}
+
+### Form Theme
+
+To render forms in a clean and user friendly way, PrestaShop extended Symfony's [Bootstrap 4 Form Theme][sf-bootstrap4-form-theme] to create **PrestaShop UI Kit Form theme**. Learn more on [Symfony Form theme for PrestaShop]({{< relref "/8/development/components/form/form-theme/form-theme">}})
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
To render forms in a clean and user-friendly way, PrestaShop extended Symfony's [Bootstrap 4 Form Theme][sf-bootstrap4-form-theme] to create **PrestaShop UI Kit Form theme**. Learn more on [Symfony Form theme for PrestaShop]({{< relref "/8/development/components/form/form-theme/form-theme">}})
```

File: development/components/form/_index.md [hunk 10]
```diff
@@ -1,8 +1,276 @@
...
+
+## Customize forms by modules
+
+### Concepts 
+
+- `FormModifier` - Allows to modify the `Form`, mostly for modules
+- `FormModifier` hook_ - Allows to connect the `FormModifier` to the Form
+- `FormDataProviderData` hook - Allows to provide data to the form
+- `FormDataProviderDefaultData` hook - Allows to provide default data to the form
+- `FormHandler hook` - Allows to handle the form and its data with modules
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
- `FormModifier`: This component is used to modify the Form, primarily for module integration.
- `FormModifier` hook_: Connects the FormModifier to the form, enabling modifications.
- `FormDataProviderData` hook: Facilitates the provision of specific data to the form.
- `FormDataProviderDefaultData` hook: Used for supplying default data to the form.
- `FormHandler hook`: Allows the module to handle the form and its data effectively.
```

File: development/components/form/_index.md [hunk 11]
```diff
@@ -1,8 +1,276 @@
...
+    end
+    G-.->ZZ(Redirect)
+    style F fill:#f9f,stroke:#333,stroke-width:2px
+    style Module fill:#dcedfc
+    style Core fill:#fbfbea
+</div>
+
+### FormModifier
+
+A `FormModifier` (or `FormBuilderModifier`) allows you to modify the content of a `Form`, in a module for example, to add, modify or remove an element.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
A `FormModifier` (also known as `FormBuilderModifier`) allows you to alter the contents of a Form. It is particularly useful within modules, allowing developers to add, modify, or remove elements from the form as required.
```

File: development/components/form/_index.md [hunk 12]
```diff
@@ -1,8 +1,276 @@
...
+    style F fill:#f9f,stroke:#333,stroke-width:2px
+    style Module fill:#dcedfc
+    style Core fill:#fbfbea
+</div>
+
+### FormModifier
+
+A `FormModifier` (or `FormBuilderModifier`) allows you to modify the content of a `Form`, in a module for example, to add, modify or remove an element. 
+
+It as been implemented in an example module: [DemoProductForm2](https://github.com/PrestaShop/example-modules/blob/master/demoproductform2/src/Form/Modifier/ProductFormModifier.php).
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
It's been implemented in an example module: [DemoProductForm2](https://github.com/PrestaShop/example-modules/blob/master/demoproductform2/src/Form/Modifier/ProductFormModifier.php). 
```

File: development/components/form/_index.md [hunk 13]
```diff
@@ -1,8 +1,276 @@
...
+    PrestaShop\Module\DemoProductForm\Form\Modifier\ProductFormModifier:
+        autowire: true
+        public: true
+        arguments:
+            $formBuilderModifier: '@form.form_builder_modifier'
+```
+
+#### FormModifier hook
+
+The `FormModifier`, without being hooked to a `Form` will have no effect. The `FormModifier` must be hooked by implementing [action&lt;Object&gt;FormBuilderModifier]({{<relref "/8/modules/concepts/hooks/list-of-hooks/action<FormName>FormBuilderModifier">}}).
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
A `FormModifier` by itself will not affect a `Form` unless it is properly hooked. To ensure functionality, the `FormModifier` must be linked by implementing the [action<Object>FormBuilderModifier]({{<relref "/8/modules/concepts/hooks/list-of-hooks/action<FormName>FormBuilderModifier">}}).
```

File: development/components/form/_index.md [hunk 14]
```diff
@@ -1,8 +1,276 @@
...
+
+In the module, register the hook and implement the method, for example, for the `Customer` entity:
```
Comment (kpodemski) [inline development/components/form/_index.md]: I think this example is for the `Product`?

File: development/components/form/_index.md [hunk 15]
```diff
@@ -1,8 +1,276 @@
...
+    $this->registerHook('actionProductFormDataProviderDefaultData');
+}
+
+public function hookActionProductFormDataProviderDefaultData(array $params): void
+{
+    // define default values for fields in $params["data"]
+}
+```
+
+This hook has been implemented as an example in our [modules examples repository - demoformdataproviders](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
This hook has been implemented as an example in our [example-modules repository](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```

File: development/components/form/_index.md [hunk 16]
```diff
@@ -1,8 +1,276 @@
...
+    $this->registerHook('actionProductFormDataProviderData');
+}
+
+public function hookActionProductFormDataProviderData(array $params): void
+{
+    // change value of fields in $params["data"]
+}
+```
+
+This hook has been implemented as an example in our [modules examples repository - demoformdataproviders](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
This hook has been implemented as an example in our [example-modules repository - demoformdataproviders](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```

File: development/components/form/_index.md [hunk 17]
```diff
@@ -1,8 +1,276 @@
...
+
+### FormHandler hook
+
+The `FormHandler` hooks allows to receive the `Form` data, and make operations with this data to persist it.
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
The `FormHandler` hook is designed to intercept `Form` data, enabling the execution of various operations with this data. Its primary function is to facilitate the persistence of form data, ensuring that the data captured by the form can be effectively processed and stored later.
```

File: development/components/form/_index.md [hunk 18]
```diff
@@ -1,8 +1,276 @@
...
+
+4 hooks are available to achieve this:
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
There are 4 hooks available to work with the `Form` data:
```

File: development/components/form/_index.md [hunk 19]
```diff
@@ -1,8 +1,276 @@
...
+public function hookActionAfterUpdateCombinationFormFormHandler(array $params): void
+{
+    $combinationId = $params['form_data']['id'];
+    [...]
+}
+```
+
+## Examples
+
+You can find some examples about how to use forms in your modules to extend PrestaShop capacities:
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
You can find some examples of how to use forms in your modules to extend PrestaShop capabilities: 
```

File: development/components/form/_index.md [hunk 20]
```diff
@@ -1,8 +1,276 @@
...
+}
+```
+
+## Examples
+
+You can find some examples about how to use forms in your modules to extend PrestaShop capacities: 
+
+| Subject | Link |
+| --- | --- |
+| Extend a Symfony form in a module with upload image field | [Tutorial]({{<relref "/8/modules/sample-modules/extending-sf-form-with-upload-image-field" >}}) |
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
| Extend a Symfony form in a module with an upload field for image | [Tutorial]({{<relref "/8/modules/sample-modules/extending-sf-form-with-upload-image-field" >}}) |
```

File: development/components/form/_index.md [hunk 21]
```diff
@@ -1,8 +1,276 @@
...
+| Extending the new product page form | [Tutorial]({{<relref "/8/modules/sample-modules/extend-product-page" >}}) |
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
| How to extend the new product page form | [Tutorial]({{<relref "/8/modules/sample-modules/extend-product-page" >}}) |
```

File: development/components/form/_index.md [hunk 22]
```diff
@@ -1,8 +1,276 @@
...
+| Add action in grid to modify customers | [Tutorial]({{<relref "/8/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage" >}}) |
```
Comment (kpodemski) [inline development/components/form/_index.md]:
```suggestion
| How to modify Customers grid | [Tutorial]({{<relref "/8/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage" >}}) |
```

File: modules/concepts/hooks/list-of-hooks/action<FormName>FormDataProviderData.md [hunk 23]
```diff
@@ -33,3 +32,7 @@ $this->hookDispatcher->dispatchWithParameters(
     ]
 );
 ```
+
+## Example implementation
+
+This hook has been implemented as an example in our [modules examples repository - demoformdataproviders](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```
Comment (kpodemski) [inline modules/concepts/hooks/list-of-hooks/action<FormName>FormDataProviderData.md]:
```suggestion
This hook has been implemented as an example in our [example-modules repository](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```

File: modules/concepts/hooks/list-of-hooks/action<FormName>FormDataProviderDefaultData.md [hunk 24]
```diff
@@ -31,3 +30,7 @@ dispatchWithParameters(
             ]
         )
 ```
+
+## Example implementation
+
+This hook has been implemented as an example in our [modules examples repository - demoformdataproviders](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```
Comment (kpodemski) [inline modules/concepts/hooks/list-of-hooks/action<FormName>FormDataProviderDefaultData.md]:
```suggestion
This hook has been implemented as an example in our [example-modules repository](https://github.com/PrestaShop/example-modules/tree/master/demoformdataproviders).
```

**Review by kpodemski — COMMENTED** [review]
Let's go! Review ready :)

I wanted to improve some wording and make some things a bit more rich with the use of different wording 👍🏻

File: development/components/form/_index.md (hunk 14 above)
Comment (thomasnares) [inline development/components/form/_index.md]:
```suggestion
In the module, register the hook and implement the method, for example, for the `Product` entity: 
```

**Comment by kpodemski** [comment]
Thanks @thomasnares 

Yet another great project 👍🏻

**Pull request merged**